crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -3455,54 +3457,63 @@ impl<'db> TypeInferenceBuilder<'db> {
                 Some(ty)
             }
 
-            (Type::Instance(_), Type::IntLiteral(_), op) => self.infer_binary_expression_type(
-                left_ty,
-                KnownClass::Int.to_instance(self.db()),
-                op,
-            ),
-
-            (Type::IntLiteral(_), Type::Instance(_), op) => self.infer_binary_expression_type(
-                KnownClass::Int.to_instance(self.db()),
-                right_ty,
-                op,
-            ),
-
-            (Type::Instance(_), Type::Tuple(_), op) => self.infer_binary_expression_type(
-                left_ty,
-                KnownClass::Tuple.to_instance(self.db()),
-                op,
-            ),
-
-            (Type::Tuple(_), Type::Instance(_), op) => self.infer_binary_expression_type(
-                KnownClass::Tuple.to_instance(self.db()),
-                right_ty,
-                op,
-            ),
-
-            (Type::Instance(_), Type::StringLiteral(_) | Type::LiteralString, op) => self
-                .infer_binary_expression_type(left_ty, KnownClass::Str.to_instance(self.db()), op),
-
-            (Type::StringLiteral(_) | Type::LiteralString, Type::Instance(_), op) => self
-                .infer_binary_expression_type(KnownClass::Str.to_instance(self.db()), right_ty, op),
+            (Type::BooleanLiteral(b1), Type::BooleanLiteral(b2), ast::Operator::BitOr) => {
+                Some(Type::BooleanLiteral(b1 | b2))
+            }
 
-            (Type::Instance(_), Type::BytesLiteral(_), op) => self.infer_binary_expression_type(
-                left_ty,
-                KnownClass::Bytes.to_instance(self.db()),
+            (Type::BooleanLiteral(bool_value), right, op) => self.infer_binary_expression_type(
+                Type::IntLiteral(i64::from(bool_value)),
+                right,
                 op,
             ),
+            (left, Type::BooleanLiteral(bool_value), op) => {
+                self.infer_binary_expression_type(left, Type::IntLiteral(i64::from(bool_value)), op)
+            }
 
-            (Type::BytesLiteral(_), Type::Instance(_), op) => self.infer_binary_expression_type(
-                KnownClass::Bytes.to_instance(self.db()),
-                right_ty,
+            // We've handled all of the special cases that we support for literals, so we need to
+            // fall back on looking for dunder methods on one of the operand types.
+            (
+                Type::FunctionLiteral(_)
+                | Type::ModuleLiteral(_)
+                | Type::ClassLiteral(_)
+                | Type::SubclassOf(_)
+                | Type::Instance(_)
+                | Type::KnownInstance(_)
+                | Type::Union(_)
+                | Type::Intersection(_)
```

Comment:
I've expanded this out into a list of all of the coercable types, just like @AlexWaygood  requested in https://github.com/astral-sh/ruff/pull/15045#discussion_r1890556626